orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,557 @@
+      return true;
+    }
+
+    return func(filename);
+  });
+}
+
+Status OrtSaveInternal(
+    const std::vector<ONNX_NAMESPACE::TensorProto>& trainable_tensor_protos,
+    const std::vector<ONNX_NAMESPACE::TensorProto>& non_trainable_tensor_protos,
+    const PathString& checkpoint_path) {
+  // Make sure name unique across trainable and non-trainable lists.
+  std::unordered_set<std::string> trainable_unique_names;
+  std::unordered_set<std::string> non_trainable_unique_names;
+  std::vector<std::string> inter_sec;
+  auto check_unique = [](const std::vector<ONNX_NAMESPACE::TensorProto>& tensor_protos,
+                         std::unordered_set<std::string>& unique_names) {
+    for (auto& tensor_proto : tensor_protos) {
+      ORT_ENFORCE(unique_names.find(tensor_proto.name()) == unique_names.end(),
+                  "Duplicated tensor proto named ", tensor_proto.name());
+      unique_names.emplace(tensor_proto.name());
+    }
+  };
+  check_unique(trainable_tensor_protos, trainable_unique_names);
+  check_unique(non_trainable_tensor_protos, non_trainable_unique_names);
+  std::set_intersection(trainable_unique_names.begin(), trainable_unique_names.end(),
+                        non_trainable_unique_names.begin(), non_trainable_unique_names.end(),
+                        std::back_inserter(inter_sec));
+  ORT_RETURN_IF_NOT(inter_sec.empty(), "Tensor name exists in both trainable param list and non-trainable param list.");
+
+  // Keep following saving logic aligned with OrtSaveModuleStatesInternal.
+  LOGS_DEFAULT(INFO)
+      << "Saving model checkpoint files to " << ToUTF8String(checkpoint_path);
+  LOGS_DEFAULT_IF(Env::Default().FolderExists(checkpoint_path), WARNING)
+      << "Checkpoint directory exists - data may be overwritten.";
+  ORT_RETURN_IF_ERROR(Env::Default().CreateFolder(checkpoint_path));
+
+  // Save TensorProto to file.
+  if (trainable_tensor_protos.size() > 0) {
+    WriteTensorProtoToFile(
+        GetTensorProtoFilePath(checkpoint_path, k_trainable_param_root_prefix),
+        trainable_tensor_protos, "[trainable_param]");
+  }
+
+  if (non_trainable_tensor_protos.size() > 0) {
+    WriteTensorProtoToFile(
+        GetTensorProtoFilePath(checkpoint_path, k_non_trainable_param_root_prefix),
+        non_trainable_tensor_protos, "[non_trainable_param]");
+  }
+
+  return Status::OK();
+}
+
+Status OrtSaveModuleStatesInternal(ModuleCheckpointState& module_state,
+                                   const PathString& parameter_folder_path) {
+  // Write weight tensors files.
+  const auto& param_states = module_state.named_parameters;
+  if (!param_states.empty()) {
+    ORT_ENFORCE(module_state.train_session_data_transfer_mgr,
+                "module checkpoint state has null train_session_data_transfer_mgr.");
+
+    std::unordered_map<PathString, std::unordered_map<std::string, OrtValue>>
+        parameter_ort_values;
+    parameter_ort_values[k_trainable_param_root_prefix] = {};
```

Comment:
nit: initializing here is unnecessary as unordered_map::operator[] will already do that